Eng 5025  evaluation set

melisa-writer opened the pull request:
### Adding scripts for test data generation. 
_Remark: I'm not sure where to upload files. It's not a part of the service library -> I've created a subfolder (added to .dockerignore)_

- **get_phrase.py** - looks for verb/noun phrases in pos tags sequences
- **get_options.py** - conjugates verbs (pyinflect lib) or looks for derivationally related forms (NLTK Wordnet)
- **get_gap.py** - uses the above to generate masked sentence, answer and options
- **main.py** - reads csv, applies get_gap to sentences and writes to json
- **utils.py** - helper functions used in main.py (removed from main for clarity)

Sample output: **world_news_in_month_tense.json**

### Updates: 
- dockerignore update (see remark)
- use multi_rank
- GPU support

melisa-writer committed:
first

melisa-writer committed:
add readme

melisa-writer committed:
Update README.md

melisa-writer committed:
arrange files

melisa-writer committed:
Update .dockerignore

melisa-writer committed:
add requirements

melisa-writer committed:
Update README.md

melisa-writer committed:
use rank_multi in fitb

melisa-writer committed:
fix link

melisa-writer committed:
add GPU support

sam-qordoba commented on fitbert_test_data/get_gap.py:
```diff
@@ -0,0 +1,105 @@
+            end = chosen_verb_phrase["end"]
+
+            options = get_verb_options(main_verb, verb_phrase_inffix)
+            options_len = [1 / len(o) for o in options]
```
Can we add some checks around here to make 100% sure we aren't dividing by 0?

melisa-qordoba replied:
There is no chance to get division by zero here. 
See: implementation of conjugate. I am checking whether VB form exists. If VB exist then there are no options of len 0 possible.

sam-qordoba commented on fitbert/fitb.py:
```diff
@@ -177,12 +160,37 @@ def simplify_options(self, sent: str, options: List[str]):
+        options = [replaced]
+        return self.fitb(masked_str, options, delemmatize=True)
+
+
```
really nit-picky, but the file should end with one blank line.

I use `black` as an auto-formatter and recommend everyone else does, it will automatically fix things like this

sam-qordoba replied:
This is what my `.vscode/settings.json` looks like (I install certain things like linters and formatters with anaconda, so I don't have to install them in every env):

```json
{
    "python.pythonPath": "venv/bin/python",
    "python.linting.enabled": true,
    "python.linting.mypyEnabled": true,
    "python.linting.mypyPath": "/anaconda3/bin/mypy",
    "python.linting.mypyArgs": [
        "--ignore-missing-imports",
        "--namespace-packages",
        "--follow-imports=silent",
        "--show-column-numbers"
    ],
    "python.linting.pylintEnabled": false,
    "python.formatting.provider": "black",
    "python.formatting.blackPath": "/anaconda3/bin/black",
    "files.exclude": {
        "**/.git": true,
        "**/.svn": true,
        "**/.hg": true,
        "**/CVS": true,
        "**/.DS_Store": true,
        "**/*.pyc": true,
        "**/__pycache__": true,
        "**/.mypy_cache": true
    }
```

melisa-qordoba replied:
Good point! I forgot to run black here. Thx!

melisa-qordoba replied:
Done.

sam-writer committed:
markdown linting fixes and add CLI option for NLTK install

remove trailing whitespace
replace hard tabs with spaces
add newlines to make linter happy
explain how to install nltk deps from CLI

sam-qordoba commented on fitbert_test_data/get_phrase.py:
```diff
@@ -0,0 +1,90 @@
+import re
+
+dictionary = {
```
is this format from wordnet?

melisa-qordoba replied:
There are 2 reasons to write custom mapping:

1. Non compatible/obsolete alternatives:

- spacy - only noun chunks

- wordnet - different than spacy pos tag system
https://www.guru99.com/pos-tagging-chunking-nltk.html
https://spacy.io/api/annotation

- textacy - obsolete as we had the following changes:
https://universaldependencies.org/v2/postags.html
CONJ -> CCONJ
In fact also all SCONJ has been mapped to CCONJ (probe on 90k corpus).

2. Possibility to add statistical gaps in the near future: 

Custom mapping allows to find longest common pos tag sequences using Sequence Matcher (https://docs.python.org/2/library/difflib.html).
-> given crowd-sourced examples of fill in the blanks we can find statistically probable places for gaps.
This piece of code has been implemented (and in fact took >75% of coding time), but since no open source examples are available atm - skipped.

sam-qordoba replied:
Can you walk me through this on a call tomorrow?

melisa-qordoba replied:
Sure.

sam-qordoba commented on fitbert_test_data/get_options.py:
```diff
@@ -0,0 +1,425 @@
+    options = []
+    diff_order = ["there", "here", "everywhere", "elsewhere", "anywhere", "nowhere"]
+    for verb_form in verb_forms:
+        if (verb_form[4] in tense) and (verb_form[5] in form):
```
I really try to not access indices directly - is there a way to make this more literate? e.g. `if is_present_perfect(verb_form):` or whatever it is you are testing?

melisa-qordoba replied:
The list has been changed to a dictionary. It looks nicer indeed.

sam-qordoba commented on fitbert_test_data/get_options.py:
```diff
@@ -0,0 +1,425 @@
+    for verb_form in verb_forms:
+        if (verb_form[4] in tense) and (verb_form[5] in form):
+            if verb_phrase_inffix.lower() in diff_order:
+                elements = [verb_phrase_inffix, verb_form[0]] + verb_form[1:3]
```
Same with this one - can the code use functions or something to make more clear what is happening here?

melisa-qordoba replied:
Changed to filters over dictionaries.

sam-qordoba commented on fitbert_test_data/README.md:
````diff
@@ -0,0 +1,71 @@
+nltk.download()
+```
+
+## Usage
````
Is there a command that we could put in this section that will work out of the box?

melisa-qordoba replied:
Nice readme fix, thx!

sam-writer reviewed (approved):
Approving, because I don't want to block the process, but I do have some things I'd like fixed

sam-qordoba commented on fitbert/fitb.py:
```diff
@@ -12,13 +12,15 @@
 class FitBert:
     def __init__(
         self,
+        device=None,
```
I want these arguments in order of descending-estimated-order-of-usage, but I really can't guess what they will be. Honestly, since there will probably be improvements to BERT in the near future (or [past](https://ai.facebook.com/blog/roberta-an-optimized-method-for-pretraining-self-supervised-nlp-systems/)), maybe we want `model_name` first? I think `device` will be near the end.

sam-qordoba commented on fitbert/fitb.py:
```diff
@@ -12,13 +12,15 @@
     ):
         self.mask_token = mask_token
         self.delemmatizer = Delemmatizer()
+        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
```
I think we should check if the user passed in `device` here: if they did, it's probably them trying to tell us to use GPU or CPU. Maybe we should check if they pass `device="cuda"` or `device="cpu"`, and respect that. If neither, check if it is truthy, in which case set `device="cuda"`, and if falsey, set `device="cpu"`?

melisa-qordoba replied:
Agreed, we can have such fix here. Although I would wait till we finalize 'insides' first, then add options etc.

sam-writer committed:
update main readme, add colab how-to

sam-qordoba commented on fitbert_test_data/utils.py:
```diff
@@ -0,0 +1,81 @@
+
+
+def is_valid_sentence(sentence):
+    return not "..." in sentence
```
Is this specific to some of the training data? Because `...` is a valid punctuation mark.

melisa-qordoba replied:
Removing incomplete sentences ending with "...." - condition specific for this corpus, although one could add "ending with "...." condition. Further checking would be required whether its enough (they might end with ... .)
or
you could use a different dataset or preprocess it before running the script.

sam-qordoba commented:
I added a few more comments, and an update to the main readme. I made a [colab notebook to show usage](https://colab.research.google.com/drive/1NHYRMR7pKt9uY_KfFhIKXwPwvuX-VQLW) and put it in the readme.

I ❤️ this PR. Thank you. I hope you know, the comments all come from a place of loving this code and wanting it to be 💅

melisa-writer committed:
change verb forms to dict

melisa-writer committed:
add black formatting

melisa-writer committed:
avoid double masking

melisa-writer committed:
add encoding options

sam-qordoba commented on fitbert_test_data/main.py:
```diff
@@ -0,0 +1,83 @@
+        doc = nlp(text)
+        for sentence in doc.sents:
+
+            if is_valid_sentence(sentence.text):
```
This might be overkill, but just to bring it up so we can talk about it on the call - can we make `is_valid_sentence` somehow a parameter to this script? Since validity checking is based on the dataset?
